src/coreclr/jit/compiler.h

```diff
@@ -6839,16 +6839,9 @@ class Compiler
     bool optExtractInitTestIncr(
         BasicBlock** pInitBlock, BasicBlock* bottom, BasicBlock* top, GenTree** ppInit, GenTree** ppTest, GenTree** ppIncr);
 
-    enum class RedirectBlockOption
-    {
-        DoNotChangePredLists, // do not modify pred lists
-        UpdatePredLists,      // add/remove to pred lists
-        AddToPredLists,       // only add to pred lists
-    };
-
     void optRedirectBlock(BasicBlock*      blk,
-                          BlockToBlockMap* redirectMap,
-                          const RedirectBlockOption = RedirectBlockOption::DoNotChangePredLists);
+                          BasicBlock*      newBlk,
+                          BlockToBlockMap* redirectMap);
```

Comment:
I'm not sure I understand the new function that well -- it requires me to have a block around that the previous function did not require me to have around, but what if I don't have that?

This function doesn't seem to be doing redirection anymore, it rather seems to be initializing a new block based on an old block and a map. Which is fine if that's what we generally need, but the old function seemed much more general than this -- basically a generalization of `fgReplaceJumpTarget` to replace multiple targets at once. How would we implement that functionality with the new scheme? It is conceivable to me that we are going to need it again.

Reply:
> it requires me to have a block around that the previous function did not require me to have around, but what if I don't have that?

For the current callers of `optRedirectBlock`, `newBlk` is obtained from `redirectMap` -- for each copied block `blk`, we expect there to be a `blk -> newBlk` mapping. I could refactor `optRedirectBlock` slightly so that it only takes `blk`, and retrieves `newBlk` from `redirectMap`. We'd always expect this mapping to exist, right? Can you foresee a situation where we wouldn't have that?

> This function doesn't seem to be doing redirection anymore, it rather seems to be initializing a new block based on an old block and a map. Which is fine if that's what we generally need, but the old function seemed much more general than this

In my opinion, the old version was trying to do a few too many things, making the semantics of its callers a bit awkward. Previously, callers would have to copy the old block's target(s) to the new block with `BasicBlock::CopyTarget` without setting up any pred edges, and then tell `optRedirectBlock` to add pred edges, while deciding what the real target should be. I don't think this approach will hold up well once we've replaced block targets with successor edges: Once we've done that, it won't be possible for a block to have a jump target without an edge to it, which is a state we'd previously establish before calling `optRedirectBlock`.

> How would we implement that functionality with the new scheme? It is conceivable to me that we are going to need it again.

This would introduce some code duplication, but maybe we could have a version of `optRedirectBlock` that assumes `newBlk`'s successors are initialized, and the current version (which we can rename to imply it is initializing successors, if you'd prefer) can continue to assume successors aren't initialized?

Reply:
> For the current callers of optRedirectBlock, newBlk is obtained from redirectMap -- for each copied block blk, we expect there to be a blk -> newBlk mapping. I could refactor optRedirectBlock slightly so that it only takes blk, and retrieves newBlk from redirectMap. We'd always expect this mapping to exist, right? Can you foresee a situation where we wouldn't have that?

I would not expect this mapping to always exist. As I mentioned above, I think `optRedirectBlock` should be viewed as a generalization of `fgReplaceJumpTarget` that allows replacing multiple targets in one operation. I.e. it was an operation similar to
```cpp
for ((BasicBlock* key, BasicBlock* target) in blockMap) { fgReplaceJumpTarget(block, key, target); }
```
I agree the overloaded handling around preds before wasn't very pretty.

I think with the refactoring done in this PR the function should be renamed to something that indicates that it operates on a partially initialized block and is some form of initialization -- it no longer matches the above semantics. Maybe something like `optInitializeDuplicatedBlockTargets`?

I actually had local changes where I generalized `optRedirectBlock` to take a functor instead of a `BlockToBlockMap`, making it more easy to use for these kinds of more general block redirections. With that I think `fgReplaceJumpTarget` could be implemented in terms of `optRedirectBlock` (with a functor like `[=](BasicBlock* target) { if (target == oldBlock) { return newBlock; } return oldBlock; }`).

Reply:
> I think with the refactoring done in this PR the function should be renamed to something that indicates that it operates on a partially initialized block and is some form of initialization -- it no longer matches the above semantics. Maybe something like optInitializeDuplicatedBlockTargets?

Sure thing, I can include that change in my next PR.

> I actually had local changes where I generalized optRedirectBlock to take a functor instead of a BlockToBlockMap, making it more easy to use for these kinds of more general block redirections.

I like that idea. Are you ok with me wrapping up the successor edge refactor first, and then coming back to this? That should only take a few more PRs; I have the code ready locally.

Reply:
Sure -- also, I'm ok with leaving the function out entirely and leaving it as it is right now (with the rename). We can add the more general version when we find a need for it.